data/data/gcp/network/common.tf

```diff
@@ -1,2 +1,24 @@
 # Canonical internal state definitions for this module.
 # read only: only locals and data source definitions allowed. No resources or module blocks in this file
+
+data "google_compute_network" "preexisting_cluster_network" {
+  count = var.preexisting_network ? 1 : 0
+
+  name = var.cluster_network
+}
```

Comment:
Do we get an error if we fetch a data block that doesn't exist? The reason I ask is because it might be good to let the outcome of this data fetch determine if preexisting_network is true (in locals) instead of passing it around. We can instead set the name(s) for these in tfvars and let terraform work out the logic as to what that means.

Reply:
> Do we get an error if we fetch a data block that doesn't exist? 

Yes. I have not found a way to safely check whether a resource exists in Terraform. This seems to be by design. As I'm reading more about Terraform and its declarative approach it seems the push should be in the opposite direction of moving conditional logic out of the configuration, so I have been looking into that.

There is still some conditional logic within the network module, so this is a compromise (that I am fine with).  If we wanted to to remove all conditional logic I think what we would do is create a separate configuration to read an existing network and one to create a new network. In gocode we would conditionally `apply` the appropriate configuration, then run `apply` a second time onthe rest of the configuration with the output from the first run. 

This would keep all of the conditional logic in gocode and the TF configs would be purely declarative with dependency inversion. The plan above is my attempt to fill in the blanks of (this doc which seemingly describes what we want)[https://www.terraform.io/docs/modules/composition.html#conditional-creation-of-objects]. 

I like the plan above in terms of pure principal but I don't see a significant benefit, so this current code is the best I have been able to come up with.